### PR TITLE
implement approval handler

### PR DESCRIPTION
**Pull request by hongchaodeng**
ref: https://github.com/kubernetes/contrib/issues/1389

What?
This PR adds ApprovalHandler that will try to add "approved" label once all files of change has been approved by approvers.

```
// The algorithm goes as:
// - Initially, we set up approverMap
//   - Go through all comments after latest commit. If any approver said "/approve", add him to approverMap.
// - For each file, we see if any approver of this file is in approverMap.
//   - An approver of a file is defined as:
//     - It's known that each dir has a list of approvers. (This might not hold true. For usability, current situation is enough.)
//     - Approver of a dir is also the approver of child dirs.
//   - We look at only top 3 dir level's approvers .
// - Iff all files has been approved, the bot will add "approved" label.
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/contrib/1727)

### CODE REVIEW

**Comment by hongchaodeng** [comment]
cc @bgrant0607 @philips @xiang90

File: mungegithub/mungers/approval-handler.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+limitations under the License.
+*/
+
+package mungers
+
+import (
+	"path"
+	"strings"
+
+	"github.com/golang/glog"
```
Comment (foxish) [inline mungegithub/mungers/approval-handler.go]: The imports need to be reordered.

File: mungegithub/mungers/approval-handler.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		return
+	}
+
+	comments, err := getCommentsAfterLastModified(obj)
+	if err != nil {
+		glog.Errorf("failed to get comments in this PR: %v", err)
+		return
+	}
+
+	approverMap := map[string]struct{}{}
```
Comment (foxish) [inline mungegithub/mungers/approval-handler.go]: If the intent is to use it like a set, we could use sets.String (`k8s.io/kubernetes/pkg/util/sets`) instead?

File: mungegithub/mungers/approval-handler.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+			return
+		}
+	}
+	obj.AddLabel(approvedLabel)
+}
+
+func (h *ApprovalHandler) hasApproval(filename string, approverMap sets.String) bool {
+	paths := strings.Split(filename, "/")
+	var p string
+	for i := 0; i < len(paths) && i < 3; i++ {
```
Comment (grodrigues3) [inline mungegithub/mungers/approval-handler.go]: can we make 3 a parameter so it's clear that it represents the maxDepth of the search?

File: mungegithub/mungers/approval-handler.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+		return
+	}
+
+	comments, err := getCommentsAfterLastModified(obj)
+	if err != nil {
+		glog.Errorf("failed to get comments in this PR: %v", err)
+		return
+	}
+
+	approverMap := sets.String{}
```
Comment (grodrigues3) [inline mungegithub/mungers/approval-handler.go]: nit: change the variable name from approverMap to approverSet if it's a set.

File: mungegithub/mungers/approval-handler.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+	if !obj.IsPR() {
+		return
+	}
+	files, err := obj.ListFiles()
+	if err != nil {
+		glog.Errorf("failed to list files in this PR: %v", err)
+		return
+	}
+
+	comments, err := getCommentsAfterLastModified(obj)
```
Comment (grodrigues3) [inline mungegithub/mungers/approval-handler.go]: @apelisse in the event that file A is modified, only file A should require re-approval right?  I think this implementation will require all approvers write ```/approve``` after every new commit, if if the file or files they previously approved have not changed at all.  is that right?

Comment (hongchaodeng) [inline mungegithub/mungers/approval-handler.go]: That's a good point. I think this is one of the optimization that we can do in the future.
I want to keep make this PR simple and usable due to lack of testing. Once it's working, we can incrementally add optimizations.

**Comment by hongchaodeng** [comment]
All comments addressed.
PTAL.

**Comment by apelisse** [comment]
/lgtm

---

Reviewed 2 of 3 files at r1, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[mungegithub/mungers/approval-handler.go, line 86 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/1727#-KRH799tXp1kBxHfpYSt-r3-86:-KRdnz8xw20YraeJuVEA:b-4au5xj) ([raw file](https://github.com/kubernetes/contrib/blob/f43f3b8fc5565415e777960b3d5b6cf1fc0010d1/mungegithub/mungers/approval-handler.go#L86)):*
<details><summary><i>Previously, hongchaodeng (Hongchao Deng) wrote…</i></summary>
> That's a good point. I think this is one of the optimization that we can do in the future.
> 
> I want to keep make this PR simple and usable due to lack of testing. Once it's working, we can incrementally add optimizations.
</details>
Yeah, that's a good point. Pretty useful optimization, but let's go for incremental :-)

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/contrib/1727#-:-KRdpa4XjSPeghNRbcJO:b-8ks5qa)*

**Pull request merged**